### DISCUSSION TITLE
Multipage real-time data visualization with multithreading

### DISCUSSION THREAD

**Discussion by mauf3r** [Avaiga/taipy#1667] [Q&A]
I started to move an app from Streamlit to Tapir, because of the limitations with regard to real-time data.
I implemented my app with the help of [this example](https://docs.taipy.io/en/latest/tutorials/visuals/5_multithreading/) wich is working basically.

Unfortunately I did not manage to invoke the data-receiving callable if the page was defined as a class.

My question:
I have multiple data-sources, running in different threads, invoked at main() level of the app.
What is the best practice for the data flow?

My first attempt was, to define the data in every sub-page and have a data-receiving callable in every sub-page. in that scenario I did not manage to exchange data between the pages, for example select one row in a df, and show a different page with the details of that data.
Next step was to define the data on the main() level, now the data exchange between pages works, but the callables on each page are trying to update the same data, wich gives me errors and warnings.

I appreciate any hints.

**Comment by FlorianJacta** [comment]
Why do you have to have callables on each page?

**Comment by mauf3r** [comment]
That' exactly, why I am asking :-)

I need all pages to update, when the data is updated. And I think I did not fully understand the 'state' yet.
So I was asking myself, if I can do all the updating on the main() level - without the need of the 'invoke_callback' method.. The pages then would get no additional trigger.
A tgb.table has the "rebuilt" option, so would it automatically refresh the display of the data frame, which ist updated elsewhere?
Why with a tgb.text element? - and  I am also using tab.part(page='some url') - this is not updated if the url changes so I have to do a 'navigate' to reload the page.

**Reply by FlorianJacta** [reply]
You have two scopes for variables in Taipy. You have global and page variables.

Global variables are shared by everyone while page variables are not. If you defined a variable in the main, this would become a global variable, and a change applied to it will update every visual element requiring this variable (don't define variables with the same name on other pages).

Also, in a callback, you can access every variable, so you could have no global variables but change every page variable within the same callback.

```python
def callback(state): 
   state['page_1'].data = ...
   state['page_2'].data = ...
   ...
```

>  and I am also using tab.part(page='some url') - this is not updated if the url changes so I have to do a 'navigate' to reload the page.

Could you extend on this part so that I can understand the issue?

**Comment by mauf3r** [comment]
I am working on a trading-bot - my data are my transaction history, signals from tradingview and account informations from the exchange.
On one page I have a part() showing a widget with a price-chart for a certain trading-pair. The pair to be shown ist either defined on a different page, followed by a navigate call to that certain page. Or, if that certain page is already active, the chart should be changed by an action on that page.

Your little example gives me the idea, that I should  have all relevant clallback-funktionality on the main() app level. That may solve all my issues.

given the following structure:
```python
with tgb.Page() as my_page:
  tgb.part(page="{my_url}")
  tgb.table("{my_df}", id='my_page_table', rebuild=True)
``` 

and in the main-app:
```python
def on_action(state, id, payload):
  if id == 'my_page_table':
    state['my_page'].my_url = some_function(payload)
    navigate(state, 'My_Page')
    
pages = {'My_Page': my_page}

Gui(pages=pages).run()
``` 

If I click on a line in that table, the url is updated according to that line. But I need to reload the page, after I changed the url - correct?

**Reply by FlorianJacta** [reply]
That might be the case. Have you tried?

What you could do if what is not updated is only in the part:

- create a [partial](https://docs.taipy.io/en/develop/manuals/userman/gui/pages/partial/) where you will show your page
- update it in place of navigate

```python
with tgb.Page() as my_page:
  tgb.part(partial="{external_page}")
  tgb.table("{my_df}", id='my_page_table', rebuild=True)
```

```python
def build_external_page():
    with tgb.Page() as page:
        tgb.part(page="{my_url}")
    return page
    

def on_action(state, id, payload):
  if id == 'my_page_table':
    state['my_page'].my_url = some_function(payload)
    state.external_page_update_content(state, build_external_page())
    
pages = {'My_Page': my_page}

gui = Gui(pages=pages)
external_page = gui.add_partial(build_external_page())
gui.run()
```

**Comment by mauf3r** [comment]
> That might be the case. Have you tried?

Yes, the part() only got updated after reloading the page.

I changed the organization of the app.
The receiver is called in the __main__ module, and updating of the data In the pages works as expected.

Now im am struggling with the on_action() callback.
As far as I understand the doc, a global on_action callback should be called by any guy element in any page as soon, as an action is recognized. Without the need, to specify the on_action attribute in the definition of the guy element.

Having the on_action defined in __main__  and no 'on_action = on_action' attribute in the tgb.table(), does not invoke the callable, when I click somewhere in the table.
Putting the  'on_action = on_action' attribute into the tgb.table() , generates an error.
Having a local on_action in the page, is also ignored, unless I define the attribute.

**Reply by FlorianJacta** [reply]
Could you provide some code? Or the message error? Can you replicate the issue with a minimal example so that we can work on it?
